docs on contexts and expressions (#3571)
Napari/napari: docs/guides/contexts_expressions.md

# Contexts and Expressions in napari

```{note}
This page is mostly aimed at developers who are interested in contributing to

> or understanding the inner workings of napari.  While the concept expressions

Reviewer: ```suggestion
or understanding the inner workings of napari.  While the concept of expressions
```

*will* be relevant for plugin developers writing plugin manifests, it won't be
necessary for plugin devs or end-users to understand the implementation details
described on this page.
```

*Contexts* and *Expressions* are two concepts being introduced along with the second-generation napari plugin engine (npe2) that capture the abstract idea of "some condition" (an `Expression`) that can be evaluated at some later time, with a concrete set of keys and values (the `Context`).

## Python expressions

> In python, **expressions** are simple combinations of **values** and **operations** that can be reduced to a single value. For example, `1 > 5` is an expression that always reduces to the value`False` when evaluated.  `x > 5 and y == 'hello'` is also an expression that reduces to a boolean value; however, in order to evaluate that expression, we need to be able to fill in the values for the variable **names** "`x`" and "`y`". Those values are provided by some **context** (or "namespace"), which maps the variable names to their values.

Reviewer: ```suggestion
In Python, **expressions** are simple combinations of **values** and **operations** that can be reduced to a single value. For example, `1 > 5` is an expression that always reduces to the value `False` when evaluated.  `x > 5 and y == 'hello'` is also an expression that reduces to a boolean value; however, in order to evaluate that expression, we need to be able to fill in the values for the variable **names** "`x`" and "`y`". Those values are provided by some **context** (or "namespace"), which maps the variable names to their values.
```

> The value of `expression` depends on the context in which it is evaluated.

Reviewer: ```suggestion
The value of an `expression` depends on the context in which it is evaluated.
```

```python
In [1]: expression = "x > 5 and y == 'hello'"

In [2]: context_a = {'x': 7, 'y': 'hello'}

In [3]: context_b = {'x': 8, 'y': 'howdie!'}

In [4]: eval(expression, context_a)
Out[4]: True

In [5]: eval(expression, context_b)
Out[5]: False
```

## Napari expressions

In napari, we'd like to be able to capture the concept of some condition being `True` or `False`, *prior* to actually having the context required to evaluate it.  For example, a plugin (or napari itself) might want to stipulate that a given function should only be enabled when "the active layer has at least 3 dimensions".

> At runtime, in python code, this might be captured by the expression:

Reviewer: ```suggestion
At runtime, in Python code, this might be captured by the expression:
```

```python
viewer.layers.selection.active.data.ndim >= 3
```

... however, if you don't have access to the future `viewer` instance, that doesn't work.  So napari has the concept of `Expr` objects that represent an expression "without a context", to be evaluated later.

```{tip}

> napari's `Expr` class subclasses from [`ast.AST`](https://docs.python.org/3/library/ast.html#ast.AST) and shares many similarities with the `body` of [`ast.Expr`](https://docs.python.org/3/library/ast.html#ast.Expr). However, for the sake of evaluation safety, they only support a subset of operations, omitting things like function calls, generators, comprehensions, and collections. It's not important to fully understand ASTs to use napari expressions, but for a good introduction to python's abstract syntax tree (AST) module, see https://greentreesnakes.readthedocs.io.

Reviewer: ```suggestion
napari's `Expr` class subclasses from [`ast.AST`](https://docs.python.org/3/library/ast.html#ast.AST) and shares many similarities with the `body` of [`ast.Expr`](https://docs.python.org/3/library/ast.html#ast.Expr). However, for the sake of evaluation safety, napari's `Expr` only supports a subset of operations, omitting things like function calls, generators, comprehensions, and collections. It's not important to fully understand ASTs to use napari expressions, but for a good introduction to Python's abstract syntax tree (AST) module, see https://greentreesnakes.readthedocs.io.
```

)
```

The expression object can be evaluated using `eval`

```python        
In [9]: expr.eval({'x': 7, 'y': 'hello'})
Out[9]: True
```

and it can also be combined with other expressions & constants using **operators**:

```python
In [10]: new_expr = expr & (7 > 10)  # always False

In [11]: new_expr.eval({'x': 7, 'y': 'hello'})
Out[12]: False
```

### napari context keys

To capture napari-specific conditions, napari will declare special string **names** that can be used in a napari expression.  Taking the example above, a plugin might only want to provide a function if “the active layer has at least 3 dimensions”. For this, napari recognizes the variable `active_layer_ndim` used in an expression. In a plugin manifest, the plugin can provide a *when clause* to enable/disable a given command:

```yaml
command:
   id: myplugin.my_command
   when: active_layer_ndim >= 3
```

Some example context key names (currently) include:

| Name     | Description |
| -------- | -------- |
| `layers_selection_count`  | Number of layers currently selected  | 
| `all_layers_linked`  | True when all selected layers are linked  | 
| `active_layer_is_rgb`  | True when the active layer is RGB  | 

> | `active_layer_type`  | Lowercase name of active layer type, or None of none active.  | 

Reviewer: ```suggestion
| `active_layer_type`  | Lowercase name of active layer type, or None if no layer is active.  | 
```

    )
    active_layer_ndim = ContextKey(
        None,
        "Number of dimensions in the active layer, or `None` if nothing is active",
        lambda s: s.active and getattr(s.active.data, "ndim", None),
    )
```

You can think of a `ContextNamespace` kind of like an `Enum` class.  And just like an `Enum`, you can see all of its members:

```python
In [13]: LayerListContextKeys.__members__
Out[13]:
mappingproxy({
    'layers_selection_count': ContextKey(id='layers_selection_count'),
    'active_layer_is_rgb': ContextKey(id='active_layer_is_rgb'),
    'active_layer_ndim': ContextKey(id='active_layer_ndim')
})
```

A nice aspect of `ContextKeys` is that they can be used in expressions.  But unlike a simple string, they can also provide type hinting, linting capabilities, and IDE autocompletion.

```python
In [14]: expr = LayerListContextKeys.active_layer_ndim >= 3 

In [15]: expr.eval({'active_layer_ndim': 2})
Out[15]: False
```

A record of all registered context keys can be retrieved with the class method `ContextKey.info()`

## Contexts

Now that we've seen how expression names are declared, let's discuss the "context" in which these expressions are evaluated.

> As mentioned, a context is ultimately just a mapping between variable names and their values. (When evaluating a napari expression with `Expr.eval`, you can indeed just pass a `dict`. )

Reviewer: ```suggestion
As mentioned, a context is ultimately just a mapping between variable names and their values. (When evaluating a napari expression with `Expr.eval`, you can indeed just pass a `dict` as that mapping.)
```

> Important objects in napari, such as the `Viewer` and the `LayerList` will be associated with a `Context` object that tracks the value of various context keys.  It is the job of these various objects (i.e. the `Viewer` and the `LayerList`) to update the values in their Contexts when they change.  Continuing with the example above, if the user clicks on a 4-dimensional layer, the layerlist would set the context key `active_layer_ndim` to `4`.  (napari would then be able to enable/disable various commands & menus that required on a specific number of dimensions in the active layer)

Reviewer: ```suggestion
Important objects in napari, such as the `Viewer` and the `LayerList` will be associated with a `Context` object that tracks the value of various context keys.  It is the job of these various objects (i.e. the `Viewer` and the `LayerList`) to update the values in their `Contexts` when they change.  Continuing with the example above, if the user clicks on a 4-dimensional layer, the `LayerList` would set the context key `active_layer_ndim` to `4`.  (napari would then be able to enable/disable various commands & menus that required a specific number of dimensions in the active layer)
```

### The `Context` class

The napari `Context` class is a subclass of [`collections.ChainMap`](https://docs.python.org/3/library/collections.html#collections.ChainMap) that also emits events when a key has been modified. `ChainMap` is useful here as it allows us to have "sub-contexts" that are children of some parent context.  Child contexts can access all of the keys of the parent (but not vice-versa).  For example because a `Viewer` has a `LayerList`, all of the keys in the `Viewer` context are available to the `LayerList` context. 

```python
In [1]: from napari.utils.context import get_context

In [2]: ctx = get_context(viewer.layers)

In [3]: ctx
Out[3]: Context(
    {
    'layers_selection_count': 0,
    'all_layers_linked': False,
    'unselected_linked_layers': 0,
    'active_layer_is_rgb': False,
    'active_layer_type': None,
    'only_images_selected': False,
    'only_labels_selected': False,
    'active_layer_ndim': None,
    'active_layer_shape': None,
    'active_layer_dtype': None,
    'all_layers_same_shape': False
    }, 
    {},  # the viewer context ... (no keys yet)
    SettingsAwareContext({})
)
```

> The "root" context is a special `SettingsAwareContext` that can access keys in the global `settings`.  Because contexts are `ChainMaps`, they can all access the settings:

Reviewer: I didn't know this but this is so cool!

```python
In [4]: ctx['settings.appearance.theme']
Out[4]: 'dark'
```

When we evaluate an expression, we can provide it one of these context objects:

```python
In [5]: expr = LayerListContextKeys.layers_selection_count > 0

In [6]: expr.eval(ctx)
Out[6]: False
```

### Updating Contexts

> You may be wondering exactly how objects such as `Viewer` and `LayerList` update the keys in their contexts. The aforementioned [`ContextNamespace`](#contextkey-objects) comes in to play here again.  A `ContextNamespace` can be instantiated, and bound to a specific `Context` instance.

Reviewer: ```suggestion
You may be wondering exactly how objects such as `Viewer` and `LayerList` update the keys in their contexts. The aforementioned [`ContextNamespace`](#contextkey-objects) comes into play here again.  A `ContextNamespace` can be instantiated, and bound to a specific `Context` instance.
```

# Contexts and Expressions in napari

> ```{note}

Reviewer 2: I like this clarifying note, and will probably do the same in any docs I write!

y == 'hello'` is also an expression that reduces to a boolean value; however, in
order to evaluate that expression, we need to be able to fill in the values for
the variable **names** "`x`" and "`y`". Those values are provided by some
**context** (or "namespace"), which maps the variable names to their values.

The value of an `expression` depends on the context in which it is evaluated.

```python
In [1]: expression = "x > 5 and y == 'hello'"

In [2]: context_a = {'x': 7, 'y': 'hello'}

In [3]: context_b = {'x': 8, 'y': 'howdie!'}

In [4]: eval(expression, context_a)
Out[4]: True

In [5]: eval(expression, context_b)
Out[5]: False
```

## Napari expressions

In napari, we'd like to be able to capture the concept of some condition being
`True` or `False`, *prior* to actually having the context required to evaluate
it.  For example, a plugin (or napari itself) might want to stipulate that a
given function should only be enabled when "the active layer has at least 3
dimensions".

At runtime, in Python code, this might be captured by the expression:

```python
viewer.layers.selection.active.data.ndim >= 3
```

> ... however, if you don't have access to the future `viewer` instance, that

Reviewer 2: ```suggestion
However, if you don't have access to the future `viewer` instance, that
```

doesn't work.  So napari has the concept of `Expr` objects that represent an

> expression "without a context", to be evaluated later.

Reviewer 3: I really like this problem statement and I wonder if it shouldn't go right at the top of the guide.

```{tip}
napari's `Expr` class subclasses from
[`ast.AST`](https://docs.python.org/3/library/ast.html#ast.AST) and shares many
similarities with the `body` of
[`ast.Expr`](https://docs.python.org/3/library/ast.html#ast.Expr). However, for
the sake of evaluation safety, napari's `Expr` only supports a subset of
operations, omitting things like function calls, generators, comprehensions, and
collections. It's not important to fully understand ASTs to use napari
expressions, but for a good introduction to Python's abstract syntax tree (AST)
module, see https://greentreesnakes.readthedocs.io.
```

A string expression can be converted to a napari expression with
`parse_expression`:

```python
In [6]: from napari.utils.context import parse_expression

In [7]: expr = parse_expression("x > 5 and y == 'hello'")

In [8]: expr
Out[8]:
BoolOp(
  op=And(),
  values=[
    Compare(
      left=Name(id='x', ctx=Load()),
      ops=[Gt()],
      comparators=[Constant(value=5)]
    ),
    Compare(
      left=Name(id='y', ctx=Load()),
      ops=[Eq()],
      comparators=[Constant(value='hello')]
    )
  ]
)
```

> 

Reviewer 3: what do you think about introducing the concept of the ast in the "python expressions" section.  Then here you can focus on the napari evaluator and lead in to the context keys discussion.

The expression object can be evaluated using `eval`

```python        
In [9]: expr.eval({'x': 7, 'y': 'hello'})
Out[9]: True
```

and it can also be combined with other expressions & constants using
**operators**:

```python

> In [10]: new_expr = expr & (7 > 10)  # always False

Reviewer 4: even if it's neat that you can do that with Python constant, I think it might be better for a tutorial to 
do 
```
In [10]: other_expr = parse_expr("7 > 10")

In [11]: new_expr = expr & other_expr
```

That does work in all the case, even when what is on the RHS of `&` is not purely operator and constants. 
Yes it's more typing and strictly unnecessary but that does not mix concepts. 

Just a though depending on the level of your audiance.

Reviewer 4: Reading it more, I think that's ok to leave the `new_expr = expr & (7 > 10)` as this is for people to understand inner napari working.

In [11]: new_expr.eval({'x': 7, 'y': 'hello'})

> Out[12]: False

Reviewer 4: In 11 ; out 12.  One of the two is off by one.

Author: hah.  caught modifying :)

```

### napari context keys

> To capture napari-specific conditions, napari will declare special string
> **names** that can be used in a napari expression.  Taking the example above, a

Reviewer 4: ```suggestion
To capture napari-specific conditions, napari will declare special
**names** that can be used in a napari expression.  Taking the example above, a
```

Maybe remove `string` ? Does it matter wether it's a string ?

plugin might only want to provide a function if “the active layer has at least 3

> dimensions”. For this, napari recognizes the variable `active_layer_ndim` used

Reviewer 4: For uniformity, you above call it a name, and here a variable. Do we want to always refer to these in the same way ?
Especially since we only have expression, we are never going to have assignments, and thus not variables ?

| `active_layer_type`  | Lowercase name of active layer type, or None if no layer is active.  | 
| `only_images_selected`  | True when there is at least one selected layer and all selected layers are images  | 
| `active_layer_ndim` | Number of dimensions in the active layer, or `None` if nothing is active |
...  many more

### `ContextKey` objects

To track the special **name** strings that can be used in expressions, napari
has the `ContextKey` class.  Instances of `ContextKey` subclass from
[`ast.Name`](https://docs.python.org/3/library/ast.html#ast.Name) and represent
a variable name in an expression. Additionally, they have a `description` that
can be used in documentation, a `default_value`, and a `getter` function that
can be called to retrieve the current value (the parameters passed to the
`getter` will depend on the context key... see [Updating
Contexts](#updating-contexts) below for more).

In order to keep related keys together, `ContextKey` instances will usually be
declared as class attributes on a `ContextNamespace` class:

```python
# all of the getters here receive an instance of viewer.layers.selection

> class LayerListContextKeys(ContextNamespace):

Reviewer 4: Hum, thinking about it I don't like the `LayerListContextKeys` name for the class. It mixes the Terminologies of Keys and Namespace. Would it be better to name it `LayerListContextNamespace` ?

Author: let's address that elsewhere.  This isn't something that anyone outside of napari will use.  easy to change later

    layers_selection_count = ContextKey(

>         default_value=0,

Reviewer 2: It would be good to describe when `default_value` is used. Is it returned when evaluating the `getter` fails?

Author: will add a note.  These are used to initialize and/or reset the respective value in the context to which they are bound.  (The `getter` actually requires an instance of some object as an input, but contexts and expressions can, exist independent of any external objects...).  Perhaps this helps explain?:

```py
In [8]: from napari.utils.context import LayerListContextKeys

In [9]: ctx = {}

In [10]: llck = LayerListContextKeys(ctx)

In [11]: ctx
Out[11]:
{
    'layers_selection_count': 0,
    'all_layers_linked': False,
    'unselected_linked_layers': 0,
    'active_layer_is_rgb': False,
    'active_layer_type': None,
    'only_images_selected': False,
    'only_labels_selected': False,
    'active_layer_ndim': None,
    'active_layer_shape': (),
    'active_layer_dtype': None,
    'all_layers_same_shape': False
}
```

        description="Number of layers currently selected",
        getter=lambda sel: len(sel),
    )
    active_layer_is_rgb = ContextKey(
        default_value=False,
        description="True when the active layer is RGB",
        getter=lambda sel: getattr(sel.active, "rgb", False),
    )
    active_layer_ndim = ContextKey(
        None,
        "Number of dimensions in the active layer, or `None` if nothing is active",
        lambda s: s.active and getattr(s.active.data, "ndim", None),
    )
```

> You can think of a `ContextNamespace` kind of like an `Enum` class.  And just

Reviewer 2: I think I'm missing something, but the members feel more like methods (i.e. functions with names and descriptions that take the same instance as their first argument), than they do enum constants.

Author: perhaps that statement isn't very useful if someone doesn't know how an `Enum` is constructed.  the `Enum` class is really just a thin shell on `EnumMeta`, and their members are definitely _not_ constants.  They are _instances_ of the Enum class on which they are declared:

```py
In [1]: from enum import Enum

In [2]: class E(Enum):
   ...:     a = 1
   ...:

In [3]: type(E.a) is E
Out[3]: True
```

(this is very much different than a typical class attribute).  So in a sense, an `Enum` class is more of a namespace for a set of related enum instances:

```py
In [4]: E.__members__
Out[4]: mappingproxy({'a': <E.a: 1>})
```

just as a `ContextNamespace` class is a namespace for a set of ContextKey instances.   perhaps the intricacies of `Enum` make this comparison potentially more confusing that useful?

Author: > (i.e. functions with names and descriptions that take the same instance as their first argument)

also, don't think of these members as functions.  Think of them as static `Expr` objects, that just happen to come with a function that, when provided an object, can "grab" the current value for their expression in some specific context

Reviewer 2: In most languages, an enum is just supposed to represent a collection of related named constants. In Python the intent is the same: https://docs.python.org/3/library/enum.html

I don't think anyone should have to understand the current implementation details of Enum in Python to understand the comparison here. So, I think removing the sentence is the best option.

Author: fair enough!

like an `Enum`, you can see all of its members:

```python
In [13]: LayerListContextKeys.__members__
Out[13]:
mappingproxy({
    'layers_selection_count': ContextKey(id='layers_selection_count'),
    'active_layer_is_rgb': ContextKey(id='active_layer_is_rgb'),
    'active_layer_ndim': ContextKey(id='active_layer_ndim')
})
```

A nice aspect of `ContextKeys` is that they can be used in expressions.  But

> unlike a simple string, they can also provide type hinting, linting

Reviewer 4: AH, I see now why you call it "string names", but I think it might not be clear for someone not already familiar with the context and expression. I'm not sure how to write that better, but I'll think about it.

Author: i agree.  i think "string names" is more confusing than it's worth.  let the comparison wait till this sentence

> capabilities, and IDE autocompletion.

Reviewer 4: Reading this I'll immediately get the questions, how to  `ContextKey`s provide type hinting ?

Author: do you think i should add information on _how_ the typing is persisted through the code?  Or do you just mean "what does that even mean"?   if the latter, it means that instances of these keys know what types they're supposed to accept and yield:

<img width="453" alt="Untitled" src="https://user-images.githubusercontent.com/1609449/140786436-c0df5ede-5051-488d-8f5d-d0ea8e08c0b2.png">

```python
In [14]: expr = LayerListContextKeys.active_layer_ndim >= 3 

In [15]: expr.eval({'active_layer_ndim': 2})
Out[15]: False
```

A record of all registered context keys can be retrieved with the class method

> `ContextKey.info()`

Reviewer 4: I think this can be ambiguous as wether it's `info()` on an instance, or `info` class method on the supper class.

## Contexts

Now that we've seen how expression names are declared, let's discuss the
"context" in which these expressions are evaluated.

As mentioned, a context is ultimately just a mapping between variable names and

> their values. (When evaluating a napari expression with `Expr.eval`, you can
> indeed just pass a `dict` as that mapping.)

Reviewer 2: ```suggestion
their values. When evaluating a napari expression with `Expr.eval`, you can
indeed just pass a `dict` as that mapping.
```

Important objects in napari, such as the `Viewer` and the `LayerList` will be
associated with a `Context` object that tracks the value of various context
keys.  It is the job of these various objects (i.e. the `Viewer` and the
`LayerList`) to update the values in their `Contexts` when they change.
Continuing with the example above, if the user clicks on a 4-dimensional layer,

> the `LayerList` would set the context key `active_layer_ndim` to `4`.  (napari
> would then be able to enable/disable various commands & menus that required a
> specific number of dimensions in the active layer)

Reviewer 2: ```suggestion
the `LayerList` would set the context key `active_layer_ndim` to `4`. napari
would then be able to enable/disable various commands and menus that required a
specific number of dimensions in the active layer.
```

    'only_labels_selected': False,
    'active_layer_ndim': None,
    'active_layer_shape': None,
    'active_layer_dtype': None,
    'all_layers_same_shape': False
    }, 
    {},  # the viewer context ... (no keys yet)
    SettingsAwareContext({})
)
```
The "root" context is a special `SettingsAwareContext` that can access keys in
the global `settings`.  Because contexts are `ChainMaps`, they can all access
the settings:

```python
In [4]: ctx['settings.appearance.theme']
Out[4]: 'dark'
```

When we evaluate an expression, we can provide it one of these context objects:

```python
In [5]: expr = LayerListContextKeys.layers_selection_count > 0

In [6]: expr.eval(ctx)
Out[6]: False
```

### Updating Contexts

You may be wondering exactly how objects such as `Viewer` and `LayerList` update
the keys in their contexts. The aforementioned
[`ContextNamespace`](#contextkey-objects) comes into play here again.  A
`ContextNamespace` can be instantiated, and bound to a specific `Context`
instance.

```python
In [6]: ctx = get_context(viewer.layers)

In [7]: llck = LayerListContextKeys(ctx)
```

Attributes of an instantiated `ContextNamespace` now act as getters (and

> setters!) of their respective `ContextKey` in the associated `Context`.

Reviewer 2: What does acting as a setter mean here? The class attributes of `ContextNamespace` are functions that can be evaluated to generate a value. Does setting mean that the function is re-evaluated and anything dependent on that value is notified of a change?

Author: > The class attributes of ContextNamespace are functions that can be evaluated to generate a value.

not quite, the class attributes are _Expr_ objects (specifically `Name` objects, similar to `ast.Name`).  as mentioned in https://github.com/napari/napari/pull/3571#discussion_r744914320 ... these are not primarily functions that can be evaluated.  The fact that they come with a function to generate a value is really more of a "feature" of the `ContextKey` class (but their core functionality is to serve as a expression object that points to a name in some context)

> What does acting as a setter mean here? 

```py
In [25]: from napari.utils.context import LayerListContextKeys

In [26]: ctx = {}

In [27]: llck = LayerListContextKeys(ctx)

In [28]: llck.unselected_linked_layers
Out[28]: 0

In [29]: ctx['unselected_linked_layers']  # the above is shorthand for this
Out[29]: 0

In [30]: llck.unselected_linked_layers = 2  # <--- it's a setter for the value in the bound context

In [31]: ctx['unselected_linked_layers']
Out[31]: 2
```

Out[12]: 1
```

## Summary: A (rough) full picture

1. napari creates special context "names" using `ContextKey` and `ContextNamespace`

    ```python
    class LayerListContextKeys(ContextNamespace):
        active_layer_type = ContextKey(
            None,
            "Lowercase name of active layer type, or None of none active.",
            lambda s: s.active and s.active._type_string
        )
        active_layer_is_rgb = ContextKey(
            False,
            "True when the active layer is RGB",
            lambda s: getattr(s.active, "rgb", False)
        )
    ```

2. _Internally_ (in napari code), we can use those objects directly to declare
   expressions in an IDE-friendly way.  For example, here we are declaratively
   populating the layer-list context menu; this is a function that will split
   the current stack into multiple layers, but it is only enabled when the
   selected image is a (non-RGB) `Image` layer.

    ```python
    'napari:split_stack': {
        'description': trans._('Split Stack'),
        'action': _split_stack,
        'enable_when': LLCK.active_layer_type == "image",
        'show_when': ~LLCK.active_layer_is_rgb,
    }
    ```

> 

Reviewer 4: I think there might be one point which might not getting through in the full document, is that this is necessary if we want to be able to update the value of the expressions when the values are changing, and why we don't use direct Python object. Maybe in the motivation at the top?
